spring-web/src/main/java/org/springframework/http/HttpHeaders.java

```diff
@@ -1623,7 +1609,25 @@ public String toString() {
 	 */
 	public static HttpHeaders readOnlyHttpHeaders(HttpHeaders headers) {
 		Assert.notNull(headers, "HttpHeaders must not be null");
-		return (headers.readOnly ? headers : new HttpHeaders(headers, true));
+		if (headers instanceof ReadOnlyHttpHeaders) {
+			return headers;
+		}
+		else {
+			return new ReadOnlyHttpHeaders(headers);
+		}
+	}
+
+	/**
+	 * Return a {@code HttpHeaders} object that can read and written to.
+	 */
+	public static HttpHeaders writableHttpHeaders(HttpHeaders headers) {
+		Assert.notNull(headers, "HttpHeaders must not be null");
+		if (headers instanceof ReadOnlyHttpHeaders) {
+			return new HttpHeaders(headers.headers);
```

Comment:
If `ReadOnlyHttpHeaders` is a wrapper around `HttpHeaders`, couldn't we check here if `headers.headers` is an instance of `HttpHeaders` and if so return it directly?

Reply:
In that case, `headers.headers` is itself a `MultiValueMap`. @poutsma and I changed `ReadOnlyHttpHeaders` so that it extends HttpHeaders (and calls `super(MultiValueMap)`) to avoid issues on the `equals` contract.

Reply:
Ok I see.